clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java

```diff
@@ -546,8 +546,9 @@ public ConsumerGroupHeartbeatRequestData buildRequestData() {
             data.setMemberEpoch(membershipManager.memberEpoch());
 
             // InstanceId - only sent if has changed since the last heartbeat
```

Comment:
For my own education? How can `group.instance.id` change? Isn't it provide by the user via a config, and configs are immutable?

Reply:
I suppose this was copied from the KIP, so it's a question of the protocol specification. I don't think in the consumer implementation it can change. It seems that either that a little imprecision in the protocol spec, or the protocol allows changing/adding the instance ID in a running client. @dajac probably knows which of the two it is.

Reply:
Might the fact that an group instance ID is required when leaving the group even be a mistake in the group coordinator?
As far as I can see, the member epoch that is sent already contains the information that a static member leaves the group and the group coordinator should keep a mapping between member ID and group instance ID. Why is   the group instance ID required when leaving?

Reply:
I was wondering that as well. I wanted to get this out of the way to stabilize the tests. It's not wrong for the client to send the instance ID here. If we are worried, we can wait for David to come back from PTO. I wouldn't want to just relax the requirements in the coordinator on my own.

Reply:
The current implementation actually requires the static member id to be set all the time if the consumer uses the static membership. So the comment in the schema was wrong. I need to go back to the implementation to see whether we could relax it and only require it in the first request, when joining. I will check and let you guys know.

Reply:
I think "setting the id all the time" vs "omitting it" is an important but orthogonal question. -- The comment say that it must be set "if has changed", but it should never change, right?

> So the comment in the schema was wrong
Is this about setting the id vs not setting it, or about the original question if it could change?

Reply:
Right. This part is clearly wrong in the doc. It cannot change within the lifetime of the consumer.

Reply:
Did a follow up PR: https://github.com/apache/kafka/pull/15519